### PR TITLE
feat(agent-server): add workspace archive endpoint for git-delta/tar.gz (AGE-1871)

### LINKED ISSUE #3779: Add agent-server endpoint for archiving workspace files
## Goal

Expose a safe agent-server endpoint that returns an archive of the active workspace files so callers can persist the agent's work before a runtime is deleted. In some cases we want to have a long-lived archive of the agent's work to prevent it from being irrecoverable.

## Scope

Add or harden an endpoint in `openhands-agent-server` for downloading a directory archive.

Relevant code path:

* `openhands-agent-server/openhands/agent_server/file_router.py`
  * Existing `/file/upload` and `/file/download` endpoints.
  * Existing `_create_zip_from_directory()` helper used for trajectory export.

## Proposed implementation

1. Add an endpoint such as `GET /file/archive?path=/workspace/project`.
2. Return a streamed `.zip` or `.tar.gz` archive of the requested directory.
3. Validate that the requested path is absolute, exists, and is a directory.
4. Prevent path traversal and avoid following symlinks outside the requested root.
5. Prefer streaming or temp-file cleanup via `BackgroundTask` so large workspaces do not remain on disk after response completion.
6. Add a manifest entry or deterministic archive metadata where practical.

## Clarification

This endpoint is the reusable agent-server primitive for workspace archive-before-delete. It should focus on archiving workspace files, not full runtime disk images. It is expected to be consumed by APP-2403, with deployment/configuration handled by APP-2405.

## Acceptance criteria

* A caller can request an archive of `/workspace/project` and receive a valid archive.
* Requests for missing paths, files instead of directories, or unsafe paths fail with clear 4xx responses.
* Symlinks do not allow files outside the requested root to be included.
* Tests cover nested files, empty directories, unsafe symlinks, missing paths, and cleanup of temporary archive files.
* The endpoint is documented in the agent-server OpenAPI output.

This update was prepared by an AI agent (OpenHands) on behalf of Graham Neubig.

### PR DESCRIPTION
**Pull request by simonrosenberg**
HUMAN:

Reviewed and verified locally — this is the agent-server primitive for archiving a workspace before a runtime is deleted, used by the cloud archive-before-delete work (PLTF-3112 / APP-2403). Closes #3779.

AGENT:

Implemented by Claude (Opus 4.8) on Simon's machine. Ran the targeted tests + ruff + pyright locally (all green).

## Why

OpenHands Cloud loses workspace state when a runtime's PVC is reaped, which blocks building evals/datasets from real conversations. This adds a safe, reusable way to pull an archive of a workspace's files (or a compact git delta) out of a running sandbox before deletion, consumed by the runtime-api pause-time path (PLTF-3112) and the app-server explicit-delete path (APP-2403).

## Summary

Adds `GET /api/file/archive` to the agent-server `file_router`:

- `format=git-delta` (default): a git patch of the working-tree delta (tracked mods, untracked additions, deletions) against `base_ref` (auto-detected: origin branch / merge-base / empty tree), built with a throwaway `GIT_INDEX_FILE` so the repo's real index is untouched. The response includes `X-Archive-Base-Commit` so consumers can replay the patch.
- `format=tar.gz`: full-file archive of the directory, with a deterministic `archive_manifest.json` (format, source, file_count, total_bytes, applied excludes). This works for non-git directories and captures gitignored/non-repo files.
- `_DEFAULT_ARCHIVE_EXCLUDES` is a default, not a hard minimum. Heavy generated/dependency dirs (`node_modules/`, `.venv/`, `__pycache__/`, `dist/`, `build/`, …) are skipped by default for compact captures, but callers can set `use_default_excludes=false` to capture fat directories/full workspaces; repeatable `exclude` patterns still allow caller-specific pruning.
- Path validation (absolute / exists / is-dir → clear 4xx) and symlink-escape safety (`os.walk(followlinks=False)` + skipping symlinked files).
- Temp archive built outside the workspace, removed via `BackgroundTask`.
- Documented in the agent-server OpenAPI output.

### REST API contract changes

Compared with base OpenAPI `217b2cf83aab` for public `/api/**` paths.

```diff
--- base public OpenAPI
+++ head public OpenAPI
@@ -27,0 +28 @@
+operation GET /api/file/archive operationId=archive_directory_api_file_archive_get
@@ -156,0 +158,5 @@
+parameter GET /api/file/archive query:base_ref required=false schema=anyOf=[type="string",type="null"]
+parameter GET /api/file/archive query:exclude required=false schema=anyOf=[type="array" items=type="string",type="null"]
+parameter GET /api/file/archive query:format required=false schema=type="string" enum=["git-delta","tar.gz"] default="git-delta"
+parameter GET /api/file/archive query:path required=true schema=type="string"
+parameter GET /api/file/archive query:use_default_excludes required=false schema=type="boolean" default=true
@@ -304,0 +311,2 @@
+response GET /api/file/archive 200 application/json schema={}
+response GET /api/file/archive 422 application/json schema=HTTPValidationError
@@ -1972,0 +1981 @@
+schema StartConversationRequest property observability_span_name optional schema=type="string" default="conversation"
```

## How to Test

```
uv run pytest tests/agent_server/test_file_router.py -v
uv run ruff check openhands-agent-server/openhands/agent_server/file_router.py tests/agent_server/test_file_router.py
uv run ruff format --check openhands-agent-server/openhands/agent_server/file_router.py tests/agent_server/test_file_router.py
```

Covers path validation, symlink-escape exclusion, temp cleanup, git-delta for fresh and committed repos, subdir scoping, invalid base refs, base commit headers, default exclude behavior, disabling default excludes, and tar.gz full-workspace archives for non-git/gitignored/fat-file cases. All 49 tests in `test_file_router.py` pass; `ruff`, `ruff format`, and `pyright` are clean.

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.13-nodejs22-slim` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.13-nodejs22-slim) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:4d9ad33-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-4d9ad33-python \
  ghcr.io/openhands/agent-server:4d9ad33-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:4d9ad33-golang-amd64
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-golang-amd64
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-golang-amd64
ghcr.io/openhands/agent-server:4d9ad33-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:4d9ad33-golang-arm64
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-golang-arm64
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-golang-arm64
ghcr.io/openhands/agent-server:4d9ad33-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:4d9ad33-java-amd64
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-java-amd64
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-java-amd64
ghcr.io/openhands/agent-server:4d9ad33-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:4d9ad33-java-arm64
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-java-arm64
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-java-arm64
ghcr.io/openhands/agent-server:4d9ad33-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:4d9ad33-python-amd64
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-python-amd64
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-python-amd64
ghcr.io/openhands/agent-server:4d9ad33-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-amd64
ghcr.io/openhands/agent-server:4d9ad33-python-arm64
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-python-arm64
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-python-arm64
ghcr.io/openhands/agent-server:4d9ad33-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-arm64
ghcr.io/openhands/agent-server:4d9ad33-golang
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-golang
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-golang
ghcr.io/openhands/agent-server:4d9ad33-golang_tag_1.21-bookworm
ghcr.io/openhands/agent-server:4d9ad33-java
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-java
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-java
ghcr.io/openhands/agent-server:4d9ad33-eclipse-temurin_tag_17-jdk
ghcr.io/openhands/agent-server:4d9ad33-python
ghcr.io/openhands/agent-server:4d9ad331b051596607b5ead4b52c5155ff08e436-python
ghcr.io/openhands/agent-server:age-1871-archive-endpoint-python
ghcr.io/openhands/agent-server:4d9ad33-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `4d9ad33-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `4d9ad33-python-amd64`) are also available if needed

### CODE REVIEW

File: openhands-agent-server/openhands/agent_server/file_router.py [hunk 1]
```diff
@@ -134,6 +145,166 @@ def _create_zip_from_directory(source_dir: Path, output_path: Path) -> None:
...
+        subprocess.run(
+            ["git", "read-tree", ref],
+            cwd=root,
+            env=env,
+            capture_output=True,
+            check=True,
+            timeout=60,
+        )
+        subprocess.run(
+            ["git", "add", "-A"],
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py]: By design — the git-delta faithfully reflects the repo's own git state and uses `git add -A`, which honors `.gitignore`. In the normal case `node_modules/` is gitignored and therefore excluded (this is exactly the "no node_modules" benefit the delta format was chosen for). The QA scenario (an untracked `node_modules/` in a repo with no `.gitignore`) is unusual; capturing it there is the correct, faithful representation of the workspace for eval replay rather than a bug. I've updated the PR description to state this precisely instead of implying an unconditional exclude. If we later want hard caps for storage, that belongs as an explicit size/exclude policy at the archive-upload layer, not by diverging from git's own view here.

[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py]: Addressed in 79ef6f5: `git-delta` now applies a default `core.excludesFile` (node_modules/, .venv/, __pycache__/, dist/, build/, …) on top of the repo's .gitignore, so an untracked, non-ignored `node_modules/` is excluded. Added a test for exactly that case.

[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py]: Fixed in 4475c61: `git add` and `git diff` now use a `-- .` pathspec (cwd=root), so a path that is a subdirectory of a larger repo yields only that subtree's delta, not the whole repo's. Added test_git_delta_scoped_to_requested_subdirectory.

File: openhands-agent-server/openhands/agent_server/file_router.py [hunk 2]
```diff
@@ -331,3 +502,99 @@ async def download_trajectory(
...
+        )
+    except GitRepositoryError as e:
+        output_path.unlink(missing_ok=True)
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail=f"Not a git repository: {e}",
+        )
+    except GitCommandError as e:
+        output_path.unlink(missing_ok=True)
+        raise HTTPException(
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py:963]: Fixed in 4475c61: an unresolvable explicit base_ref from get_valid_ref is now raised as ValueError and mapped to a 400 by the route, instead of a 500. Added test_git_delta_invalid_base_ref_returns_400.

File: openhands-agent-server/openhands/agent_server/file_router.py [hunk 3]
```diff
@@ -331,3 +535,105 @@ async def download_trajectory(
...
         background=BackgroundTask(temp_file.unlink),
     )
+
+
+@file_router.get("/archive")
+async def archive_directory(
+    path: Annotated[
+        str, Query(description="Absolute path of the directory to archive")
+    ],
+    format: Annotated[
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py]: Fixed in b0b8a96: the route now uses `archive_format` internally with `Query(alias="format")`, so the public query parameter stays `format` without shadowing the built-in.

File: openhands-agent-server/openhands/agent_server/file_router.py [hunk 4]
```diff
@@ -331,3 +535,105 @@ async def download_trajectory(
...
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="Directory not found",
+        )
+    if not target.is_dir():
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Path is not a directory",
+        )
+    if base_ref is not None and base_ref.startswith("-"):
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py:905]: Leaving this as the documented behavior intentionally: `base_ref` is accepted but ignored for `tar.gz`, which lets runtime/app-server callers reuse the same archive config across formats. The OpenAPI description says it is only used for `git-delta`, and `test_tar_gz_ignores_base_ref_without_error` covers the behavior. Also, `zip` is no longer a supported format in this PR.

File: openhands-agent-server/openhands/agent_server/file_router.py [hunk 5]
```diff
@@ -134,6 +147,343 @@ def _create_zip_from_directory(source_dir: Path, output_path: Path) -> None:
...
+
+
+def _resolve_git_repo_root(target: Path, max_depth: int = 3) -> Path:
+    """Resolve the git work-tree to archive under ``target``.
+
+    Callers pass the workspace base (e.g. ``/workspace/project``), but a
+    repository-backed conversation clones into a subdirectory
+    (``{base}/[{group}/]{repo_name}``), so the base itself is usually *not* a git
+    repo. ``git rev-parse`` only searches upward, so archiving the base would
+    400 even though the repo sits one or two levels below it — silently
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py:235]: Fixed in 9988bd8: docstring now says "depth-first" to match `frontier.pop()` (LIFO). Kept the traversal as-is — the unique/ambiguous outcome is order-independent, so DFS vs BFS doesn't change the result.

File: openhands-agent-server/openhands/agent_server/file_router.py [hunk 6]
```diff
@@ -134,6 +147,343 @@ def _create_zip_from_directory(source_dir: Path, output_path: Path) -> None:
...
+    ".pytest_cache/",
+    ".ruff_cache/",
+    "dist/",
+    "build/",
+    ".next/",
+    "target/",
+    "*.pyc",
+)
+
+# Credential-bearing git internals that must NEVER be persisted to a shared
```
[the remark this answers is not in the document]
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py:189]: Added in 9988bd8: `test_tar_gz_strips_credentials_file_and_reflog_with_tokens` builds a synthetic `.git/` by hand (so `.git/credentials` and `.git/logs/` — which `git init` doesn't write — are exercised), does a full capture (`use_default_excludes=false`), and asserts those files are absent from the members AND that none of the three distinct tokens appear in the streamed bytes, while a non-credential `.git/HEAD` still survives. (The pre-existing `test_tar_gz_*_strips_git_credentials` tests already covered `.git/config` token bytes; this closes credentials + reflog.)

**Comment by enyst** [comment]
@openhands Do a /codereview-roasted on this PR. Post your feedback as a review here with gh api.

File: openhands-agent-server/openhands/agent_server/file_router.py [hunk 7]
```diff
@@ -134,6 +147,343 @@ def _create_zip_from_directory(source_dir: Path, output_path: Path) -> None:
...
+# archive bucket, even when the caller disables the default excludes for a full
+# capture (``use_default_excludes=false``). A repo cloned with a tokenized
+# remote (``https://x-access-token:TOKEN@github.com/...``) keeps that token in
+# ``.git/config``; reflogs and saved credentials are equally sensitive. These
+# are stripped from every tar.gz regardless of the exclude settings. (git-delta
+# never includes ``.git`` — it is a working-tree diff — so this only matters for
+# the full-archive format.)
+_ALWAYS_EXCLUDE_TAR = (
+    ".git/config",
+    ".git/credentials",
```
Comment (enyst) [inline openhands-agent-server/openhands/agent_server/file_router.py]: This denylist still misses `.git/FETCH_HEAD`, which Git can populate with the exact fetched URL. With tokenized HTTPS remotes, `use_default_excludes=false` can still archive credentials into the long-lived bucket. Please either keep `.git/` excluded even for full capture, or expand the always-excluded git metadata to cover `FETCH_HEAD` and nested config locations such as `.git/modules/*/config`, with regression tests that extract the tar member and assert the token is absent.

This inline review comment was created by an AI agent (OpenHands) on behalf of the requester.

**Review by enyst — COMMENTED** [review]
🟡 **Taste Rating: Acceptable** — the shape is pragmatic and the test coverage is solid, but the `.git` credential stripping still has a real hole.

**[CRITICAL ISSUES]**
- [openhands-agent-server/openhands/agent_server/file_router.py, Line 198] **Security**: `use_default_excludes=false` still archives `.git/FETCH_HEAD`, and that file can contain the fetch URL exactly as Git saw it. For tokenized HTTPS remotes, that means the long-lived archive can still contain credentials even though the code explicitly promises to never persist credential-bearing git internals. I reproduced this locally by calling `_create_tar_gz_archive(..., excludes=[])` with a synthetic `.git/FETCH_HEAD` containing `https://x-access-token:ghs_FETCHTOK@github.com/o/r`; the tar member `project/.git/FETCH_HEAD` was included and extracted with the token intact. Submodule git dirs (`.git/modules/*/config`) are the same class of problem. The current allow-most-of-`.git` approach is too optimistic; either keep excluding `.git/` for tar.gz even in full-capture mode, or maintain a much broader denylist for all git admin files that can store remotes/credentials (`FETCH_HEAD`, nested module configs, credentials/logs/config variants, etc.) with tests that extract and assert the token is absent.

**[TESTING GAPS]**
- The added credential tests cover `.git/config`, `.git/credentials`, and `.git/logs/`, but not `.git/FETCH_HEAD` or nested submodule config files. Add regression coverage for those before trusting this archive in a shared, retained bucket.

**[RISK ASSESSMENT]**
- [Overall PR] ⚠️ Risk Assessment: 🔴 HIGH
This endpoint packages workspace contents for long-lived storage right before runtime deletion. The default path is reasonably safe, but the explicit full-capture option can still leak real Git credentials from common git metadata. That is a concrete data-exposure risk, not a theoretical edge case. **Recommendation**: do not auto-merge until the credential-denylist/full-capture policy is tightened and covered by tests.

**VERDICT:**
❌ **Needs rework**: the endpoint is useful, but the full-capture `.git` handling still leaks credentials. I’m submitting this as a COMMENT review per repo policy rather than REQUEST_CHANGES.

**KEY INSIGHT:**
The hard part is not creating the archive; it is defining a safe `.git` policy, and the current partial denylist is not safe enough for retained artifacts.

This review was created by an AI agent (OpenHands) on behalf of the requester.

---

> **Improve this review?** If any feedback above seems incorrect or irrelevant to this repository, you can teach the reviewer to do better:
>
> 1. Add a `.agents/skills/custom-codereview-guide.md` file to your branch (or edit it if one already exists) with the `/codereview` trigger and the context the reviewer is missing (e.g., "Security concerns about X do not apply here because Y"). See the [customization docs](https://docs.openhands.dev/openhands/usage/use-cases/code-review#customization) for the required frontmatter format.
> 2. Re-request a review - the reviewer reads guidelines from the PR branch, so your changes take effect immediately.
> 3. When your PR is merged, the guideline file goes through normal code review by repository maintainers.
>
> **Resolve with AI?** Install the [iterate skill](https://github.com/OpenHands/extensions/tree/main/skills/iterate) in your agent and run `/iterate` to automatically drive this PR through CI, review, and QA until it's merge-ready.
>
> Was this review helpful? React with 👍 or 👎 to give feedback.

File: openhands-agent-server/openhands/agent_server/file_router.py (hunk 7 above)
Comment (simonrosenberg) [reply openhands-agent-server/openhands/agent_server/file_router.py]: Fixed — code in `3ac11e01`, regression test in `c8fa1dd`.

Replaced the fixed-path `_ALWAYS_EXCLUDE_TAR` tuple with a position-based predicate `_is_sensitive_git_internal`: it strips `config`, `config.worktree`, `credentials`, `.git-credentials`, `FETCH_HEAD`, and every reflog under a `logs/` subtree — matched inside *any* `.git` directory, so it covers the superproject **and** submodules at `.git/modules/<name>/...` at any depth, not just the top level. A full capture still keeps the rest of `.git` for fidelity, minus those credential-bearing files.

Added `test_tar_gz_strips_fetch_head_and_submodule_credentials`: a hand-built `.git/` with token-bearing `.git/FETCH_HEAD` and `.git/modules/libfoo/{config,FETCH_HEAD,logs/HEAD}`, runs a `use_default_excludes=false` capture, extracts the tar members, and asserts none of the four distinct tokens appear in the streamed bytes while a non-credential `.git/HEAD` survives.

**Review by neubig — APPROVED** [review]
Seems reasonable to me, thank you!

**Review by enyst — APPROVED** [review]

**Comment by simonrosenberg** [comment]
@openhands fix pre-commit errors
